orangecontrib/bio/genialis/genapi.py

```diff
@@ -0,0 +1,190 @@
+from genesis import Genesis
+from genesis import GenData
+import json
+from Orange.data import *
+from .. import dicty
+import io
+import gzip
+
+DEFAULT_EMAIL = 'anonymous@genialis.com'
+DEFAULT_PASSWD = 'anonymous'
+DEFAULT_URL = 'https://dictyexpress.research.bcm.edu'
+CallBack = dicty.CallBack
+
+
+class Genapi(object):
+
+    """
+
+    Python module that leverages Genesis PyAPI (Python API for accsess to DictyExpress database).
+    It supports connection to the server and data retrieval functionalities.
+    """
+
+    def __init__(self, email=DEFAULT_EMAIL, password=DEFAULT_PASSWD, url=DEFAULT_URL, buffer=None):
+
+        self._gen = Genesis(email, password, url)
+        self.buffer = buffer
+        self.email = email
+
+    def fetch_etc_objects(self, callback=None):
+        """
+        Function downloads all available :obj:`GenData` etc objects from DictyExpress database.
+        Objects are stored in buffer for further use.
+        If buffer is set to None function returns a list of :obj:`GenData` objects.
+        """
+
+        def optcb():
+            if callback:
+                callback()
+
+        cbc = CallBack(1, optcb, callbacks=70)
+
+        store_experiments = []
+
+        list_of_experiments = self._gen.api.data.get(case_ids__contains='5535115cfad58d5e03006217', status='done',
+                                    type__startswith='data:etc:')['objects']
+        cbc.end()
+
+        cbc = CallBack(len(list_of_experiments), optcb, callbacks=30)
+
+        for dict in list_of_experiments:
+            experiment = GenData(dict, self._gen)  # create GenData object
+            store_experiments.append(experiment)
+            cbc()
+
+        cbc.end()
+
+
+        if self.buffer is not None:
+            self.to_buffer(self.email, store_experiments)  # store GenData objects into buffer
+        else:
+            cbc.end()
+            return store_experiments
+
+
+
+    def download_etc_data(self, genData,  callback=None):
+
+        """
+        Function downloads etc data of a chosen experiment from the server. Downloaded data is stored in cache.
+        If data is already in cache than it returns cached data.
+
+        :param genData: :obj:`GenData` object.
+        :return: Returns data in json like format
+        :rtype: :obj:`dict`
+        """
+
+        def optcb():
+            if callback:
+                callback()
+
+        cbc = CallBack(2, optcb, callbacks=50)
+
+        #  If experiment id is in cache, use cached data.
+        if self.buffer is not None:
+            if self.in_buffer(genData.id):
+                cbc.end()
+                return self.from_buffer(genData.id)
+
+        response = genData.download('output.etcfile')
+        cbc()
+        response_gzipped = io.BytesIO(response.content)
+        response_content = io.TextIOWrapper(gzip.GzipFile(fileobj=response_gzipped), encoding="utf-8")
+        cbc()
+        json_object = json.load(response_content)
```

Comment:
Think where you will handle errors. You will have to raise some errors here and handle them the widget that calls this as well. So many things can go wrong. HTTP statuses 400, 403, 500 and other bad responses should raise Errors. Only 200 should continue forward. The `json.load` may fail if the response text is not JSON.